Porting stock_picking_package_preparation* to 9

eLBati opened the pull request:
Modules
- stock_picking_package_preparation
- stock_picking_package_preparation_line

eLBati committed:
porting last changes from 8.0

eLBati committed:
copying stock_picking_package_preparation_line from 8.0

eLBati committed:
Removing parts related to product.ul as this model does not exist anymore
We can only compute net weight, because package weight does not exist

eLBati committed:
remove lot from operation tree, as it is a o2m now

eLBati committed:
using new method check_tracking

eLBati committed:
adapting to new values format from write method

FIX correctly handle write tuples

eLBati committed:
location_id and location_dest_id are mandatory in picking

eLBati committed:
using notebook as position based on string is not allowed

eLBati committed:
FIX tests

eLBati committed:
Adding my name as I ported modules to 9

udpate module version

stock_picking_package_preparation: IMP README and use short copyright header

OpenCode committed:
[IMP] Package Preparation Lines compatible with a multi company configuration

OpenCode committed:
[FIX] Abstract class of config class and unused import

OpenCode committed:
[IMP] Manage lot directly from ddt details

Conflicts:
	stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py

[FIX] Removed useles comment

OpenCode committed:
[IMP] Select lot on package preparation detail only for selected product

[IMP] Show/hide lot field on picking package detail based on permission

OpenCode committed:
[IMP] Add test for lot on package preparation detail

[FIX] Flake8 in test

archetipo committed:
Fix security rule  multicompany

Conflicts:
	stock_picking_package_preparation_line/README.rst

gurneyalex commented:
rebuilding on runbot.

@eLBati did you check #212 and #226 ?

gurneyalex commented:
travis should be green once #227 is merged.

jgrandguillaume commented:
LGTM :+1:

eLBati committed:
Moving pack preparation under 'inventory control' as 'operations' is only available in debug mode

eLBati committed:
Porting translations from 8.0

stock_picking_package_preparation_line: porting translations form 8.0

eLBati commented:
Hi @OpenCode 
I included your last changes to `stock_picking_package_preparation_line` to this PR.
But local tests give me

```
2016-05-13 15:44:35,913 28110 ERROR demo_9 openerp.addons.stock_picking_package_preparation_line.tests.test_package_preparation_line: `   File "/home/elbati/workspace/odoo/instances/demo9-community/parts/stock-logistics-workflow/stock_picking_package_preparation_line/tests/test_package_preparation_line.py", line 176, in test_standard_flow_with_detail_with_lot
2016-05-13 15:44:35,913 28110 ERROR demo_9 openerp.addons.stock_picking_package_preparation_line.tests.test_package_preparation_line: `     self.preparation.pack_operation_ids[0].lot_id.id,
2016-05-13 15:44:35,913 28110 ERROR demo_9 openerp.addons.stock_picking_package_preparation_line.tests.test_package_preparation_line: ` AttributeError: 'stock.pack.operation' object has no attribute 'lot_id'
```

I think this is because 9.0 introduced `stock.pack.operation.lot` model.
Could you give me any hint?

Thanks

eLBati committed:
FIX correctly handle lots on preparation lines

OpenCode commented:
@eLBati I don't know. I don't use Odoo 9 and I don't know it very well. Sorry.

eLBati commented:
@OpenCode no problem.

Lots handling and related tests should be fixed with https://github.com/OCA/stock-logistics-workflow/pull/199/commits/2def0c34a72e77efe9099efaa1f75f978b5457cf

eLBati commented:
Hello @gurneyalex , why the 'needs fixing' label?

gurneyalex commented:
@eLBati runbot was red. I've triggered a rebuild.

gurneyalex commented:
You need to update the version numbers in the **openerp**.py files (s/8.0/9.0/)

eLBati commented:
@gurneyalex thanks, done

gurneyalex commented:
@eLBati still can't figure out why runbot is red :crying_cat_face:

eLBati commented:
@gurneyalex just rebuilded runbot but `stock_picking_package_preparation` is not installable.
How can I help?

gurneyalex commented:
@eLBati what I just checked in #254 is that if your branch is merged with 9.0 then the CI tests are green. This can happen since the runbot does not merge, it really tests the branch so if you fork from a red build, you can get a red build on the PR. 

I'm :+1: on this, and I think we can move on and merge this.

lasley commented on stock_picking_package_preparation/model/stock_picking_package_preparation.py:
```diff
@@ -107,18 +102,15 @@ def _default_company_id(self):
+        string='All Content',
     )
 
     @api.one
```
api.one is deprecated, please remove all instances of it for api.multi

lasley commented on stock_picking_package_preparation/security/package_preparation_security.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
```
`<odoo noupdate="1">` & can remove data tag

lasley commented on stock_picking_package_preparation_line/README.rst:
```diff
@@ -0,0 +1,61 @@
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/
```
Please use new ReadMe template https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

lasley commented on stock_picking_package_preparation_line/__init__.py:
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+##############################################################################
```
Please replace the old headers with the [new short one](https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3)

lasley commented on stock_picking_package_preparation_line/models/ir_config.py:
```diff
@@ -0,0 +1,35 @@
+from openerp import models, fields
+
+
+class ResCompany(models.Model):
```
Please separate model by file

lasley commented on stock_picking_package_preparation_line/models/ir_config.py:
```diff
@@ -0,0 +1,35 @@
+        string='Default Picking Type used in package preparation')
+
+
+class StockPickingPackagePreparationLineSettings(models.TransientModel):
```
Class name should be UpperCamelCase version of model.name

lasley commented on stock_picking_package_preparation_line/models/ir_config.py:
```diff
@@ -0,0 +1,35 @@
+        'stock.picking.type',
+        related='company_id.default_picking_type_for_package_preparation_id')
+
+    def default_get(self, cr, uid, fields, context=None):
```
Please use new api

lasley commented on stock_picking_package_preparation_line/models/stock.py:
```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
```
Please name this file `stock_move`

lasley commented on stock_picking_package_preparation_line/models/stock.py:
```diff
@@ -0,0 +1,53 @@
+class StockMove(models.Model):
+    _inherit = 'stock.move'
+
+    def get_packs(self):
```
`@api.model`

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+    _order = 'sequence asc'
+
+    package_preparation_id = fields.Many2one(
+        'stock.picking.package.preparation', string='Stock Move',
```
required?

Also, this string attr duplicates `move_id` string. On purpose?

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+        'stock.picking.package.preparation', string='Stock Move',
+        ondelete='cascade')
+    name = fields.Text(string='Description', required=True)
+    move_id = fields.Many2one('stock.move', string='Stock Move',
```
required?

eLBati replied:
It's not, because `stock.picking.package.preparation.line` can be descriptive, without being linked to a stock move

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+        help="If you change this quantity for a 'ready' picking, the system "
+             "will not generate a back order, but will just deliver the new "
+             "quantity")
+    product_uom = fields.Many2one('product.uom')
```
`product_uom_id` - all relation fields should end in _id or _ids, depending on whether they are x2m (_ids) or x2o (_id).

This also needs a string attribute

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+                self.name = name[0][1]
+            self.product_uom = self.product_id.uom_id.id
+
+    def _prepare_lines_from_pickings(self, picking_ids):
```
`@api.model`

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+            }
+
+
+class StockPickingPackagePreparation(models.Model):
```
Please either move this into its own file (preferred), or name the file `stock_picking_package_preparation.py`

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+        pack = super(StockPickingPackagePreparation, self).create(values)
+        return pack
+
+    def _update_line_vals(self, values):
```
`api.model`

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+        return pack
+
+    def _update_line_vals(self, values):
+        # ----- Create a PackagePreparationLine for every stock move
```
Please use Python docblock styling instead of C style comments: `"""` vs `#`

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+
+    @api.multi
+    def write(self, values):
+        # ----- Delete package preparation line if the relative picking is
```
Please use Python docblock styling instead of C style comments: `"""` vs `#`

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+        if values.get('picking_ids', False):
+            package_preparation_line_model = self.env[
+                'stock.picking.package.preparation.line']
+            # ----- Collect new pickings to read the change
```
Are all the `-----`s necessary? They dirty the code IMO

OpenCode replied:
This is my style to comment :)

lasley replied:
Understood - I won't block over this, but I recommend removing. They add nothing to the code other than more maintenance. 

I am blocking on the Doc block comments though. Doc blocks are described in [PEP8](https://www.python.org/dev/peps/pep-0008/#documentation-strings) & [PEP257](https://www.python.org/dev/peps/pep-0257/)

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+
+    @api.multi
+    def action_put_in_pack(self):
+        # ----- Create a stock move and stock picking related with
```
Please use Python docblock styling instead of C style comments: `"""` vs `#`

lasley commented on stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py:
```diff
@@ -0,0 +1,280 @@
+                picking.action_confirm()
+                # ----- Show an error if a picking is not confirmed
+                if picking.state != 'confirmed':
+                    raise exceptions.Warning(
```
Warning should be renamed as UserError per pylint-odoo. Admittedly don't know the reasoning, but I assume a planned change in the future

```
R8101   Import Warning should be renamed as UserError from openerp.exceptions import Warning as UserError
```

lasley commented on stock_picking_package_preparation_line/security/package_preparation_line_security.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
```
`<odoo noupdate="1">` & can remove data tag

lasley commented on stock_picking_package_preparation_line/views/ir_config_view.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
```
`<odoo>` & remove data tag

lasley commented on stock_picking_package_preparation_line/views/stock_picking_package_preparation_line.xml:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
```
`<odoo>` & remove data tag

lasley reviewed (changes requested):
Thanks for the submission @eLBati - comments inline.

Please also rebase onto the current 9.0. If I am understanding the comments correctly, this should clear up the Runbot 🔴

eLBati committed:
IMP stock_picking_package_preparation: avoid @api.one

eLBati committed:
IMP stock_picking_package_preparation: use odoo noupdate

IMP stock_picking_package_preparation: use odoo noupdate
FIX PEP8

eLBati committed:
REF stock_picking_package_preparation_line: default_get with new API

eLBati committed:
stock_picking_package_preparation_line: FIX headers, file and model names

eLBati committed:
stock_picking_package_preparation_line: FIX ensure_one and required field

eLBati committed:
stock_picking_package_preparation_line: rename product_uom_id

FIX stock_picking_package_preparation_line : Field `product_uom` does not exist

FIX stock_picking_package_preparation_line: AttributeError: 'stock.move' object has no attribute 'product_uom_id'

eLBati committed:
IMP stock_picking_package_preparation_line comments

eLBati committed:
FIX stock_picking_package_preparation: compatibility with delivery and its ensure_one

archetipo commented on stock_picking_package_preparation/model/stock_picking_package_preparation.py:
```diff
@@ -104,7 +104,8 @@ def _default_company_id(self):
         for preparation in self:
             package = preparation.package_id
-            quants = preparation.env['stock.quant'].browse(package.get_content())
+            quants = preparation.env['stock.quant'].browse(
```
@eLBati 
package.get_content()  return a recordsets, is possible to do:
preparation.quant_ids = package.get_content()
?

eLBati replied:
@archetipo in 9.0 `get_content` returns a list of IDs

eLBati commented:
@lasley thanks, I made the changes

lasley commented:
No clue what's up with Runbot

```
stock_picking_package_preparation_line: Unmet dependencies: stock_picking_package_preparation
```

But stock_picking_package_preparation is definitely set to installable 😑 

Code looks good though 👍

lasley commented:
Once we get another signoff on this (particularly the 🔴  runbot), would you mind squashing the logical commits too please?

pedrobaeza commented:
Maybe it's a runbot glitch... Let's merge and see the final result, but please first squash a bit the commits in a logical way.

eLBati commented:
@lasley @pedrobaeza rebased and squashed
